code/datums/ai/dog/dog_behaviors.dm

```diff
@@ -0,0 +1,107 @@
+/datum/ai_behavior/battle_screech/dog
+	screeches = list("barks","howls")
+
+// Fetching makes the pawn chase after whatever it's targeting and pick it up when it's in range, with the dog_equip behavior
+/datum/ai_behavior/dog_fetch
+	behavior_flags = AI_BEHAVIOR_REQUIRE_MOVEMENT
+
+/datum/ai_behavior/dog_fetch/perform(delta_time, datum/ai_controller/controller)
+	var/mob/living/living_pawn = controller.pawn
+	var/obj/item/fetch_thing = controller.blackboard[BB_DOG_FETCH_TARGET]
+
+	if(fetch_thing.anchored || !isturf(fetch_thing.loc)) //Can't pick it up, so stop trying.
+		finish_action(controller, FALSE)
+		return
+
+	if(in_range(living_pawn, fetch_thing))
+		finish_action(controller, TRUE)
+		return
+
+	finish_action(controller, FALSE)
+
+/datum/ai_behavior/dog_fetch/finish_action(datum/ai_controller/controller, success)
+	. = ..()
+
+	if(!success) //Don't try again on this item if we failed
```

Comment:
finish_action should be for clean-up, not for extra behavior, this should be in perform()